src/jdk.compiler/share/classes/com/sun/tools/javac/util/Log.java

```diff
@@ -113,9 +134,92 @@ protected DiagnosticHandler() {
+                if (pos != null && category.annotationSuppression) {        // we should apply the Lint from the warning's position
+                    if ((lint = lintFor(diag)) == null) {
+                        addLintWaiter(currentSourceFile(), diag);           // ...but we don't know it yet, so defer
+                        return;
+                    }
+                } else                                                      // we should apply the root Lint
+                    lint = rootLint();
+            }
+            reportWithLint(diag, lint);
+        }
+
+        /**
+         * Step 2: Handle a diagnostic for which the applicable Lint instance (if any) is known and provided.
+         */
+        public final void reportWithLint(JCDiagnostic diag, Lint lint) {
+
+            // Apply hackery for REQUIRES_TRANSITIVE_AUTOMATIC (see also Check.checkModuleRequires())
+            if (diag.getCode().equals(RequiresTransitiveAutomatic.key()) && !lint.isEnabled(REQUIRES_TRANSITIVE_AUTOMATIC)) {
+                reportWithLint(
+                  diags.warning(null, diag.getDiagnosticSource(), diag.getDiagnosticPosition(), RequiresAutomatic), lint);
+                return;
+            }
+
+            // Apply the lint configuration (if any) and discard the warning if it gets filtered out
+            if (lint != null) {
+                LintCategory category = diag.getLintCategory();
+                boolean emit = !diag.isFlagSet(DEFAULT_ENABLED) ?       // is the warning not enabled by default?
+                  lint.isEnabled(category) :                            // then emit if the category is enabled
+                  category.annotationSuppression ?                      // else emit if the category is not suppressed, where
+                    !lint.isSuppressed(category) :                      // ...suppression happens via @SuppressWarnings
+                    !options.isLintDisabled(category);                  // ...suppression happens via -Xlint:-category
+                if (!emit)
+                    return;
+            }
+
+            // Proceed
+            reportReady(diag);
+        }
+
+        /**
+         * Step 3: Handle a diagnostic to which the applicable Lint instance (if any) has been applied.
+         */
+        protected abstract void reportReady(JCDiagnostic diag);
+
+        protected void addLintWaiter(JavaFileObject sourceFile, JCDiagnostic diagnostic) {
+            lintWaitersMap.computeIfAbsent(sourceFile, s -> new LinkedList<>()).add(diagnostic);
+        }
+
+        /**
+         * Flush any lint waiters whose {@link Lint} configurations are now known.
          */
-        public abstract void report(JCDiagnostic diag);
+        public void flushLintWaiters() {
+            lintWaitersMap.entrySet().removeIf(entry -> {
+
+                // Is the source file no longer recognized? If so, discard warnings (e.g., this can happen with JShell)
```

Comment:
Just for information, do you know/recall what's the exact path to get here with JShell?